### PR TITLE
Enable/Disable protection

### PR DESCRIPTION
**Pull request by hiaga**
## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/SqlVirtualMachine/docs/Remove-AzAvailabilityGroupListener.md [hunk 1]
```diff
@@ -62,6 +62,21 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -ContainerName
```
Comment (hiaga) [inline src/SqlVirtualMachine/docs/Remove-AzAvailabilityGroupListener.md]: undo this file

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/test/Disable-AzRecoveryServicesConfigureBackup.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,21 @@
+if(($null -eq $TestName) -or ($TestName -contains 'Disable-AzRecoveryServicesConfigureBackup'))
+{
+  $loadEnvPath = Join-Path $PSScriptRoot 'loadEnv.ps1'
+  if (-Not (Test-Path -Path $loadEnvPath)) {
+      $loadEnvPath = Join-Path $PSScriptRoot '..\loadEnv.ps1'
+  }
+  . ($loadEnvPath)
+  $TestRecordingFile = Join-Path $PSScriptRoot 'Disable-AzRecoveryServicesConfigureBackup.Recording.json'
+  $currentPath = $PSScriptRoot
+  while(-not $mockingPath) {
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/test/Disable-AzRecoveryServicesConfigureBackup.Tests.ps1]: remove this file

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/test/Enable-AzRecoveryServicesConfigureBackup.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,21 @@
+if(($null -eq $TestName) -or ($TestName -contains 'Enable-AzRecoveryServicesConfigureBackup'))
+{
+  $loadEnvPath = Join-Path $PSScriptRoot 'loadEnv.ps1'
+  if (-Not (Test-Path -Path $loadEnvPath)) {
+      $loadEnvPath = Join-Path $PSScriptRoot '..\loadEnv.ps1'
+  }
+  . ($loadEnvPath)
+  $TestRecordingFile = Join-Path $PSScriptRoot 'Enable-AzRecoveryServicesConfigureBackup.Recording.json'
+  $currentPath = $PSScriptRoot
+  while(-not $mockingPath) {
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/test/Enable-AzRecoveryServicesConfigureBackup.Tests.ps1]: remove this file

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/test/Enable-AzRecoveryServicesProtection.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  while(-not $mockingPath) {
+      $mockingPath = Get-ChildItem -Path $currentPath -Recurse -Include 'HttpPipelineMocking.ps1' -File
+      $currentPath = Split-Path -Path $currentPath -Parent
+  }
+  . ($mockingPath | Select-Object -First 1).FullName
+}
+
+Describe 'Enable-AzRecoveryServicesProtection' {
+    It 'ModifyRetentionPolicy' -skip {
+        { throw [System.NotImplementedException] } | Should -Not -Throw
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/test/Enable-AzRecoveryServicesProtection.Tests.ps1]: add test case

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/examples/Enable-AzRecoveryServicesProtection.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+```powershell
+{{ Add code here }}
+```
+
+```output
+{{ Add output here }}
+```
+
+{{ Add description here }}
+
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/examples/Enable-AzRecoveryServicesProtection.md:26]: add example

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/examples/Disable-AzRecoveryServicesProtection.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
+```
+
+```output
+{{ Add output here }}
+```
+
+{{ Add description here }}
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/examples/Disable-AzRecoveryServicesProtection.md]: add examples

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Helpers/backuphelper.ps1 [hunk 7]
```diff
@@ -0,0 +1,18 @@
+function Get-containerName {
+	param(
+		[System.String]$Id
+	)	
+	$containerName = (($Id -split "protectionContainers/")[1] -split "/")[0]
+	return $containerName
+
+}
+
+function Get-ItemName {
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Helpers/backuphelper.ps1]: can we rename like GetProtectableItemNameFromArmId and GetContainerNameFromArmId

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Helpers/PolicyHelpers.ps1 [hunk 8]
```diff
@@ -593,19 +592,18 @@ function ValidateMandatoryFields {
 			         $errormsg = "Weekly retention days of the week must not be empty"
 			         throw $errormsg
                  }
-                 if($Policy.RetentionPolicy.WeeklySchedule.RetentionDuration.Count -eq ($null -or 0)) 
+                 if(($Policy.RetentionPolicy.WeeklySchedule.RetentionDuration.Count -eq 0) -or ($Policy.RetentionPolicy.WeeklySchedule.RetentionDuration.Count -eq $null))
                  {
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Helpers/PolicyHelpers.ps1:596]: why we changing this ?

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1 [hunk 9]
```diff
@@ -0,0 +1,101 @@
+
+
+function Disable-AzRecoveryServicesProtection {
+	[OutputType('Microsoft.Azure.PowerShell.Cmdlets.RecoveryServices.Models.Api20230201.IProtectedItem')]
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1]: what will be the output type for RemoveRecoveryPoints ?

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1 [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        ${RetainRecoveryPointsForever}   
+    )
+
+    process
+    {   
+        if($DatasourceType -eq "AzureVM")
+        {
+            $Object=[Microsoft.Azure.PowerShell.Cmdlets.RecoveryServices.Models.Api20230201.AzureIaaSvmProtectedItem]::new() 
+        }
+        elseif($DatasourceType -eq "SAPHANA")
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1:49]: can we make this workload agnostic ?

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1 [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        {
+            $Object=[Microsoft.Azure.PowerShell.Cmdlets.RecoveryServices.Models.Api20230201.AzureVMWorkloadSapHanaDatabaseProtectedItem]::new()
+        }
+
+        $vaultName=$vaultName
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1]: should this be ?

$vaultName=$**V**aultName

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1 [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        $resourceGroupName=$ResourceGroupName
+        $itemName=$Item.Name
+        $fabricName="Azure"
+        $containerName=Get-containerName -Id $Item.Id
+
+        if($RemoveRecoveryPoints)
+        {   #disable immutability
+            #call the delete api 
+            Remove-AzRecoveryServicesProtectedItem -ContainerName $containerName -FabricName $fabricName -Name $itemName -ResourceGroupName $resourceGroupName -VaultName $vaultName
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1]: don't we have NoWait here ?

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1 [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+            if($DatasourceType -eq "AzureVM")
+            {
+                $ItemObject.ProtectedItemType="AzureIaaSVMProtectedItem"
+            }
+            elseif ($DatasourceType -eq "SAPHANA")
+            {
+                $ItemObject.ProtectedItemType="AzureVmWorkloadProtectedItem"
+            }
+        
+            New-AzRecoveryServicesProtectedItem -ContainerName $containerName -FabricName $fabricName -Name $itemName -ResourceGroupName $resourceGroupName -VaultName $vaultName -Parameter $ItemObject
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1]: can we have the NoWait functionality

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Enable-AzRecoveryServicesConfigureBackup.ps1 [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+        {
+            if($DatasourceType -eq "AzureVM")
+            {
+                $ProtectableItem= Get-AzRecoveryServicesBackupProtectableItem -ResourceGroupName $resourceGroupName -VaultName $vaultName -Filter "backupManagementType eq 'AzureIaasVM' and WorkloadType -eq 'AzureVM'" | Where-Object { $_.friendlyName -match $VMName}
+            }
+            elseif($DatasourceType -eq "SAPHANA")
+            {
+                $ProtectableItem= Get-AzRecoveryServicesBackupProtectableItem -ResourceGroupName $resourceGroupName -VaultName $vaultName -Filter "backupManagementType eq 'AzureWorkload' and WorkloadType -eq 'SAPHanaDatabase'" | Where-Object { $_.friendlyName -match $VMName}
+            }
+            $containerName=Get-containerName -Id $ProtectableItem.Id
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Enable-AzRecoveryServicesConfigureBackup.ps1]: what if $ProtectableItem is $null ?

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Enable-AzRecoveryServicesConfigureBackup.ps1 [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+        }
+        elseif($DatasourceType -eq "SAPHANA")
+        {
+            $Object=[Microsoft.Azure.PowerShell.Cmdlets.RecoveryServices.Models.Api20230201.AzureVMWorkloadSapHanaDatabaseProtectedItem]::new()
+        }
+
+        $vaultName=$VaultName
+        $resourceGroupName=$ResourceGroupName
+        $fabricName="Azure"
+        if(-not($item))   #configure backup
```
Comment (hiaga) [inline src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Enable-AzRecoveryServicesConfigureBackup.ps1]: is it $Item or $item ? have you tested this ?

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Helpers/PolicyHelpers.ps1 (hunk 8 above)
Comment (jain-arohi) [reply src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Helpers/PolicyHelpers.ps1:596]: previous one wasn't functional

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1 (hunk 9 above)
Comment (jain-arohi) [reply src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1]: there is no output for it

File: src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1 (hunk 10 above)
Comment (jain-arohi) [reply src/RecoveryServices/RecoveryServices.Backup.Autorest/custom/Cmdlets/Backup/Disable-AzRecoveryServicesConfigureBackup.ps1:49]: All azurevm, saphana and mssql have separate things to pass. even if we make it workload agnostic we have to write separate cases for mssql and saphana.

**Pull request merged**